File: src/sprintDataCollector.js
/**
 * Sprint Data Collector
 * Gathers and organizes all sprint-related data from Jira
 */

export class SprintDataCollector {
  constructor(jiraClient) {
    this.jiraClient = jiraClient;
  }

  /**
   * Collect sprint data for a specific team across projects
   */
  async collectSprintDataForTeam(projectKeys, teamLabel) {
    console.log(`Fetching sprint data for team ${teamLabel} across projects: ${projectKeys.join(', ')}...`);

    // Get the last closed sprint for this team across all projects
    const { sprint, boardId, boardName, projectKey } = await this.jiraClient.getLastClosedSprintForTeam(projectKeys, teamLabel);
    console.log(`Found closed sprint: ${sprint.name} (${projectKey} - ${boardName})`);

    // Get all issues in the sprint (filtered by team label if specified)
    const issues = await this.jiraClient.getSprintIssues(sprint.id, teamLabel);
    console.log(`Found ${issues.length} issues in sprint for team ${teamLabel}`);

    // Get team members
    const teamMembers = await this.jiraClient.getSprintTeamMembers(issues);
    console.log(`Found ${teamMembers.length} team members`);

    // Get project info
    const project = await this.jiraClient.getProject(projectKey);

    return {
      sprint,
      boardId,
      boardName,
      issues,
      teamMembers,
      project,
      projectKey,
      teamLabel,
    };
  }

  /**
   * Collect sprint data for a specific project and team
   */
  async collectSprintData(projectKey, teamLabel = null) {
    const teamInfo = teamLabel ? ` for team ${teamLabel}` : '';
    console.log(`Fetching sprint data from project ${projectKey}${teamInfo}...`);

    // Get the last closed sprint (filtered by team name if provided)
    const { sprint, boardId } = await this.jiraClient.getLastClosedSprint(projectKey, teamLabel);
    console.log(`Found closed sprint: ${sprint.name}`);

    // Get all issues in the sprint (filtered by team label if specified)
    const issues = await this.jiraClient.getSprintIssues(sprint.id, teamLabel);
    console.log(`Found ${issues.length} issues in sprint${teamInfo}`);

    // Get team members
    const teamMembers = await this.jiraClient.getSprintTeamMembers(issues);
    console.log(`Found ${teamMembers.length} team members`);

    // Get project info
    const project = await this.jiraClient.getProject(projectKey);

    return {
      sprint,
      boardId,
      issues,
      teamMembers,
      project,
      projectKey,
      teamLabel,
    };
  }

  /**
   * Collect sprint data for multiple teams across multiple projects
   */
  async collectAllSprintData(projectKeys, teamLabels) {
    const allSprintData = [];

    // If no team labels specified, discover them from all sprints
    let teamsToProcess = teamLabels;
    if (!teamLabels || teamLabels.length === 0) {
      const discoveredLabels = new Set();
      for (const projectKey of projectKeys) {
        try {
          const boards = await this.jiraClient.getBoards(projectKey);
          for (const board of boards) {
            try {
              const sprints = await this.jiraClient.getSprints(board.id);
              const closedSprints = sprints.filter(s => s.state === 'closed');
              if (closedSprints.length > 0) {
                const labels = await this.jiraClient.getTeamLabelsFromSprint(closedSprints[0].id);
                labels.forEach(label => discoveredLabels.add(label));
              }
            } catch (error) {
              console.warn(`Warning: Could not fetch sprints for board ${board.name}`);
            }
          }
        } catch (error) {
          console.warn(`Warning: Could not process project ${projectKey}`);
        }
      }
      teamsToProcess = Array.from(discoveredLabels);
      console.log(`Discovered ${teamsToProcess.length} team labels across all projects: ${teamsToProcess.join(', ')}`);
    }

    // Process each team - find their sprint across all projects
    for (const teamLabel of teamsToProcess) {
      try {
        const sprintData = await this.collectSprintDataForTeam(projectKeys, teamLabel);
        if (sprintData.issues.length > 0) {
          allSprintData.push(sprintData);
        } else {
          console.log(`No issues found for team ${teamLabel}, skipping...`);
        }
      } catch (error) {
        console.warn(`Warning: Could not collect data for team ${teamLabel}: ${error.message}`);
      }
    }

    return allSprintData;
  }

  /**
   * Calculate sprint metrics
   */
  calculateMetrics(sprintData) {
    const { issues, sprint } = sprintData;

    // Status categorization
    const statusGroups = {
      completed: [],
      inProgress: [],
      todo: [],
      blocked: [],
    };

    // Story points and issue type tracking
    let totalStoryPoints = 0;
    let completedStoryPoints = 0;
    const issueTypes = {};
    const priorities = {};

    issues.forEach(issue => {
      const status = issue.fields.status.statusCategory.name.toLowerCase();
      // Try multiple common story points field IDs
      const storyPoints = issue.fields.customfield_20826 || issue.fields.customfield_10016 || 0;
      const issueType = issue.fields.issuetype.name;
      const priority = issue.fields.priority?.name || 'None';

      // Track by status
      if (status === 'done') {
        statusGroups.completed.push(issue);
        completedStoryPoints += storyPoints;
      } else if (status === 'in progress' || status === 'indeterminate') {
        statusGroups.inProgress.push(issue);
      } else {
        statusGroups.todo.push(issue);
      }

      // Check for blockers
      if (issue.fields.labels?.includes('blocked') ||
          issue.fields.status.name.toLowerCase().includes('block')) {
        statusGroups.blocked.push(issue);
      }

      totalStoryPoints += storyPoints;

      // Track issue types
      issueTypes[issueType] = (issueTypes[issueType] || 0) + 1;

      // Track priorities
      priorities[priority] = (priorities[priority] || 0) + 1;
    });

    // Calculate dates and duration
    const startDate = new Date(sprint.startDate);
    const endDate = new Date(sprint.endDate);
    const durationDays = Math.ceil((endDate - startDate) / (1000 * 60 * 60 * 24));

    // Calculate velocity based on issues (not story points)
    const velocity = statusGroups.completed.length;
    const velocityPercentage = issues.length > 0
      ? ((statusGroups.completed.length / issues.length) * 100).toFixed(1)
      : 0;

    // Calculate completion rate
    const completionRate = issues.length > 0
      ? ((statusGroups.completed.length / issues.length) * 100).toFixed(1)
      : 0;

    return {
      totalIssues: issues.length,
      completedIssues: statusGroups.completed.length,
      inProgressIssues: statusGroups.inProgress.length,
      todoIssues: statusGroups.todo.length,
      issuesByType: issueTypes,
      blockedIssues: statusGroups.blocked.length,
      totalStoryPoints,
      completedStoryPoints,
      velocity,
      velocityPercentage: parseFloat(velocityPercentage),
      completionRate: parseFloat(completionRate),
      durationDays,
      startDate: sprint.startDate,
      endDate: sprint.endDate,
      issueTypes,
      priorities,
      statusGroups,
    };
  }

  /**
   * Analyze sprint health
   */
  analyzeSprintHealth(metrics) {
    const healthIndicators = [];
    let overallHealth = 'Good';

    // Velocity check
    if (metrics.velocityPercentage < 60) {
      healthIndicators.push({
        indicator: 'Velocity',
        status: 'Poor',
        message: `Only ${metrics.velocityPercentage}% of story points completed`,
      });
      overallHealth = 'Poor';
    } else if (metrics.velocityPercentage < 80) {
      healthIndicators.push({
        indicator: 'Velocity',
        status: 'Fair',
        message: `${metrics.velocityPercentage}% of story points completed`,
      });
      if (overallHealth === 'Good') overallHealth = 'Fair';
    } else {
      healthIndicators.push({
        indicator: 'Velocity',
        status: 'Good',
        message: `${metrics.velocityPercentage}% of story points completed`,
      });
    }

    // Completion rate check
    if (metrics.completionRate < 60) {
      healthIndicators.push({
        indicator: 'Completion Rate',
        status: 'Poor',
        message: `Only ${metrics.completionRate}% of issues completed`,
      });
      overallHealth = 'Poor';
    } else if (metrics.completionRate < 80) {
      healthIndicators.push({
        indicator: 'Completion Rate',
        status: 'Fair',
        message: `${metrics.completionRate}% of issues completed`,
      });
      if (overallHealth === 'Good') overallHealth = 'Fair';
    } else {
      healthIndicators.push({
        indicator: 'Completion Rate',
        status: 'Good',
        message: `${metrics.completionRate}% of issues completed`,
      });
    }

    // Blocked issues check
    if (metrics.blockedIssues > 0) {
      healthIndicators.push({
        indicator: 'Blockers',
        status: 'Warning',
        message: `${metrics.blockedIssues} blocked issue(s) detected`,
      });
      if (overallHealth === 'Good') overallHealth = 'Fair';
    }

    // In-progress issues check
    if (metrics.inProgressIssues > metrics.completedIssues) {
      healthIndicators.push({
        indicator: 'Work in Progress',
        status: 'Warning',
        message: 'More issues in progress than completed',
      });
    }

    return {
      overallHealth,
      healthIndicators,
    };
  }

  /**
   * Extract key accomplishments
   */
  extractAccomplishments(metrics, issues) {
    const completedIssues = metrics.statusGroups.completed;

    // Sort by priority and story points
    const sortedCompleted = completedIssues
      .sort((a, b) => {
        const priorityOrder = { 'Highest': 0, 'High': 1, 'Medium': 2, 'Low': 3, 'Lowest': 4 };
        const aPriority = priorityOrder[a.fields.priority?.name] || 5;
        const bPriority = priorityOrder[b.fields.priority?.name] || 5;
        return aPriority - bPriority;
      })
      .slice(0, 10); // Top 10 accomplishments

    return sortedCompleted.map(issue => ({
      key: issue.key,
      summary: issue.fields.summary,
      type: issue.fields.issuetype.name,
      priority: issue.fields.priority?.name || 'None',
      assignee: issue.fields.assignee?.displayName || 'Unassigned',
      storyPoints: issue.fields.customfield_10016 || 0,
    }));
  }

  /**
   * Extract current blockers
   */
  extractBlockers(metrics) {
    return metrics.statusGroups.blocked.map(issue => ({
      key: issue.key,
      summary: issue.fields.summary,
      type: issue.fields.issuetype.name,
      priority: issue.fields.priority?.name || 'None',
      assignee: issue.fields.assignee?.displayName || 'Unassigned',
      status: issue.fields.status.name,
    }));
  }

  /**
   * Generate recommendations
   */
  generateRecommendations(metrics, healthAnalysis) {
    const recommendations = [];

    if (metrics.velocityPercentage < 80) {
      recommendations.push({
        category: 'Velocity',
        priority: 'High',
        recommendation: 'Consider reducing sprint commitment or identifying impediments affecting team velocity',
      });
    }

    if (metrics.blockedIssues > 0) {
      recommendations.push({
        category: 'Blockers',
        priority: 'High',
        recommendation: `Address ${metrics.blockedIssues} blocked issue(s) immediately to prevent future sprint delays`,
      });
    }

    if (metrics.inProgressIssues > metrics.completedIssues) {
      recommendations.push({
        category: 'WIP Limit',
        priority: 'Medium',
        recommendation: 'Too much work in progress. Consider implementing WIP limits to improve flow',
      });
    }

    if (metrics.todoIssues > 0) {
      recommendations.push({
        category: 'Sprint Planning',
        priority: 'Medium',
        recommendation: `${metrics.todoIssues} issue(s) not started. Review sprint planning and capacity`,
      });
    }

    if (recommendations.length === 0) {
      recommendations.push({
        category: 'General',
        priority: 'Low',
        recommendation: 'Sprint executed well. Continue current practices and look for incremental improvements',
      });
    }

    return recommendations;
  }
}

File: sprint_summary_agent/sprint_data_collector.py
"""Sprint Data Collector - Gathers and organizes sprint-related data from Jira."""

from datetime import datetime
from typing import Any, Dict, List, Optional

from .jira_client import JiraClient


class SprintDataCollector:
    """Gathers and organizes all sprint-related data from Jira."""

    def __init__(self, jira_client: JiraClient):
        """Initialize with Jira client."""
        self.jira_client = jira_client

    def collect_sprint_data_for_team(
        self, project_keys: List[str], team_label: str
    ) -> Dict[str, Any]:
        """Collect sprint data for a specific team across projects."""
        print(f"Fetching sprint data for team {team_label} across projects: {', '.join(project_keys)}...")

        # Get the last closed sprint for this team across all projects
        result = self.jira_client.get_last_closed_sprint_for_team(project_keys, team_label)
        sprint = result["sprint"]
        board_id = result["boardId"]
        board_name = result["boardName"]
        project_key = result["projectKey"]

        print(f"Found closed sprint: {sprint['name']} ({project_key} - {board_name})")

        # Get all issues in the sprint (filtered by team label)
        issues = self.jira_client.get_sprint_issues(sprint["id"], team_label)
        print(f"Found {len(issues)} issues in sprint for team {team_label}")

        # Get team members
        team_members = self.jira_client.get_sprint_team_members(issues)
        print(f"Found {len(team_members)} team members")

        # Get project info
        project = self.jira_client.get_project(project_key)

        return {
            "sprint": sprint,
            "boardId": board_id,
            "boardName": board_name,
            "issues": issues,
            "teamMembers": team_members,
            "project": project,
            "projectKey": project_key,
            "teamLabel": team_label,
        }

    def collect_sprint_data(
        self, project_key: str, team_label: Optional[str] = None
    ) -> Dict[str, Any]:
        """Collect sprint data for a specific project and team."""
        team_info = f" for team {team_label}" if team_label else ""
        print(f"Fetching sprint data from project {project_key}{team_info}...")

        # Get the last closed sprint
        result = self.jira_client.get_last_closed_sprint(project_key, team_label)
        sprint = result["sprint"]
        board_id = result["boardId"]

        print(f"Found closed sprint: {sprint['name']}")

        # Get all issues in the sprint
        issues = self.jira_client.get_sprint_issues(sprint["id"], team_label)
        print(f"Found {len(issues)} issues in sprint{team_info}")

        # Get team members
        team_members = self.jira_client.get_sprint_team_members(issues)
        print(f"Found {len(team_members)} team members")

        # Get project info
        project = self.jira_client.get_project(project_key)

        return {
            "sprint": sprint,
            "boardId": board_id,
            "issues": issues,
            "teamMembers": team_members,
            "project": project,
            "projectKey": project_key,
            "teamLabel": team_label,
        }

    def collect_all_sprint_data(
        self, project_keys: List[str], team_labels: List[str]
    ) -> List[Dict[str, Any]]:
        """Collect sprint data for multiple teams across multiple projects."""
        all_sprint_data = []

        # If no team labels specified, just get the latest sprint for each project
        if not team_labels:
            print("No team labels provided - fetching latest sprint for each project without team filtering")

            for project_key in project_keys:
                try:
                    sprint_data = self.collect_sprint_data(project_key, None)
                    if sprint_data["issues"]:
                        all_sprint_data.append(sprint_data)
                    else:
                        print(f"No issues found in latest sprint for project {project_key}, skipping...")
                except Exception as e:
                    print(f"Warning: Could not collect data for project {project_key}: {e}")

            return all_sprint_data

        # Process each team - find their sprint across all projects
        for team_label in team_labels:
            try:
                sprint_data = self.collect_sprint_data_for_team(project_keys, team_label)
                if sprint_data["issues"]:
                    all_sprint_data.append(sprint_data)
                else:
                    print(f"No issues found for team {team_label}, skipping...")
            except Exception as e:
                print(f"Warning: Could not collect data for team {team_label}: {e}")

        return all_sprint_data

    def calculate_metrics(self, sprint_data: Dict[str, Any]) -> Dict[str, Any]:
        """Calculate sprint metrics."""
        issues = sprint_data["issues"]
        sprint = sprint_data["sprint"]

        # Status categorization
        status_groups = {
            "completed": [],
            "inProgress": [],
            "todo": [],
            "blocked": [],
        }

        # Story points and issue type tracking
        total_story_points = 0
        completed_story_points = 0
        issue_types = {}
        priorities = {}

        for issue in issues:
            fields = issue.get("fields", {})
            status = fields.get("status", {}).get("statusCategory", {}).get("name", "").lower()

            # Try multiple common story points field IDs
            story_points = fields.get("customfield_20826") or fields.get("customfield_10016") or 0

            issue_type = fields.get("issuetype", {}).get("name", "Unknown")
            priority = fields.get("priority", {}).get("name") if fields.get("priority") else "None"

            # Track by status
            if status == "done":
                status_groups["completed"].append(issue)
                completed_story_points += story_points
            elif status in ("in progress", "indeterminate"):
                status_groups["inProgress"].append(issue)
            else:
                status_groups["todo"].append(issue)

            # Check for blockers
            labels = fields.get("labels", [])
            status_name = fields.get("status", {}).get("name", "").lower()
            if "blocked" in labels or "block" in status_name:
                status_groups["blocked"].append(issue)

            total_story_points += story_points

            # Track issue types
            issue_types[issue_type] = issue_types.get(issue_type, 0) + 1

            # Track priorities
            priorities[priority] = priorities.get(priority, 0) + 1

        # Calculate dates and duration
        start_date = datetime.fromisoformat(sprint["startDate"].replace("Z", "+00:00"))
        end_date = datetime.fromisoformat(sprint["endDate"].replace("Z", "+00:00"))
        duration_days = (end_date - start_date).days

        # Calculate velocity based on issues (not story points)
        velocity = len(status_groups["completed"])
        velocity_percentage = (
            (len(status_groups["completed"]) / len(issues) * 100) if issues else 0
        )

        # Calculate completion rate
        completion_rate = (len(status_groups["completed"]) / len(issues) * 100) if issues else 0

        return {
            "totalIssues": len(issues),
            "completedIssues": len(status_groups["completed"]),
            "inProgressIssues": len(status_groups["inProgress"]),
            "todoIssues": len(status_groups["todo"]),
            "blockedIssues": len(status_groups["blocked"]),
            "totalStoryPoints": total_story_points,
            "completedStoryPoints": completed_story_points,
            "velocity": velocity,
            "velocityPercentage": round(velocity_percentage, 1),
            "completionRate": round(completion_rate, 1),
            "durationDays": duration_days,
            "startDate": sprint["startDate"],
            "endDate": sprint["endDate"],
            "issueTypes": issue_types,
            "priorities": priorities,
            "statusGroups": status_groups,
        }

    def analyze_sprint_health(self, metrics: Dict[str, Any]) -> Dict[str, Any]:
        """Analyze sprint health."""
        health_indicators = []
        overall_health = "Good"

        # Velocity check
        velocity_pct = metrics["velocityPercentage"]
        if velocity_pct < 60:
            health_indicators.append({
                "indicator": "Velocity",
                "status": "Poor",
                "message": f"Only {velocity_pct}% of story points completed",
            })
            overall_health = "Poor"
        elif velocity_pct < 80:
            health_indicators.append({
                "indicator": "Velocity",
                "status": "Fair",
                "message": f"{velocity_pct}% of story points completed",
            })
            if overall_health == "Good":
                overall_health = "Fair"
        else:
            health_indicators.append({
                "indicator": "Velocity",
                "status": "Good",
                "message": f"{velocity_pct}% of story points completed",
            })

        # Completion rate check
        completion_rate = metrics["completionRate"]
        if completion_rate < 60:
            health_indicators.append({
                "indicator": "Completion Rate",
                "status": "Poor",
                "message": f"Only {completion_rate}% of issues completed",
            })
            overall_health = "Poor"
        elif completion_rate < 80:
            health_indicators.append({
                "indicator": "Completion Rate",
                "status": "Fair",
                "message": f"{completion_rate}% of issues completed",
            })
            if overall_health == "Good":
                overall_health = "Fair"
        else:
            health_indicators.append({
                "indicator": "Completion Rate",
                "status": "Good",
                "message": f"{completion_rate}% of issues completed",
            })

        # Blocked issues check
        if metrics["blockedIssues"] > 0:
            health_indicators.append({
                "indicator": "Blockers",
                "status": "Warning",
                "message": f"{metrics['blockedIssues']} blocked issue(s) detected",
            })
            if overall_health == "Good":
                overall_health = "Fair"

        # In-progress issues check
        if metrics["inProgressIssues"] > metrics["completedIssues"]:
            health_indicators.append({
                "indicator": "Work in Progress",
                "status": "Warning",
                "message": "More issues in progress than completed",
            })

        return {
            "overallHealth": overall_health,
            "healthIndicators": health_indicators,
        }

    def extract_accomplishments(
        self, metrics: Dict[str, Any], issues: List[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        """Extract key accomplishments."""
        completed_issues = metrics["statusGroups"]["completed"]

        # Sort by priority
        priority_order = {"Highest": 0, "High": 1, "Medium": 2, "Low": 3, "Lowest": 4}

        sorted_completed = sorted(
            completed_issues,
            key=lambda x: priority_order.get(
                x.get("fields", {}).get("priority", {}).get("name"), 5
            ),
        )[:10]

        accomplishments = []
        for issue in sorted_completed:
            fields = issue.get("fields", {})
            accomplishments.append({
                "key": issue.get("key"),
                "summary": fields.get("summary", ""),
                "type": fields.get("issuetype", {}).get("name", "Unknown"),
                "priority": fields.get("priority", {}).get("name") if fields.get("priority") else "None",
                "assignee": fields.get("assignee", {}).get("displayName", "Unassigned") if fields.get("assignee") else "Unassigned",
                "storyPoints": fields.get("customfield_10016") or fields.get("customfield_20826") or 0,
            })

        return accomplishments

    def extract_blockers(self, metrics: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Extract current blockers."""
        blockers = []
        for issue in metrics["statusGroups"]["blocked"]:
            fields = issue.get("fields", {})
            blockers.append({
                "key": issue.get("key"),
                "summary": fields.get("summary", ""),
                "type": fields.get("issuetype", {}).get("name", "Unknown"),
                "priority": fields.get("priority", {}).get("name") if fields.get("priority") else "None",
                "assignee": fields.get("assignee", {}).get("displayName", "Unassigned") if fields.get("assignee") else "Unassigned",
                "status": fields.get("status", {}).get("name", "Unknown"),
            })

        return blockers
